File: scripts/mcp_telemetry_query.py
"""MCP telemetry query CLI — Phase 2 K2.

Reads the SQLite store written by ``scripts/mcp_telemetry_store.py``
and surfaces:

- Per-tool attempt counts.
- Distinct-consumer counts (``client_id_hash``).
- Outcome ratios (``implemented`` / ``stub`` / ``latent_demand``).
- Latent-demand names — tool names not in the catalog.

Refresh cadence: cheap enough to run on every ``task mcp:report``
invocation. Stdlib-only; reads (never writes) the DB.
"""
from __future__ import annotations

import argparse
import json
import sqlite3
import sys
from dataclasses import dataclass
from pathlib import Path
from typing import Any

sys.path.insert(0, str(Path(__file__).resolve().parent))
from mcp_telemetry_store import resolve_db  # noqa: E402

# Lazy import the catalog so the query CLI stays usable when only the
# JSONL is present (e.g. an analyst node without the full package).
try:
    from mcp_server.catalog import load_catalog  # noqa: E402

    _CATALOG_AVAILABLE = True
except Exception:  # pragma: no cover - defensive
    _CATALOG_AVAILABLE = False


@dataclass(frozen=True)
class ToolRow:
    tool_name: str
    attempts: int
    distinct_consumers: int
    implemented: int
    stub: int
    latent_demand: int
    last_ts: str | None

    def as_dict(self) -> dict[str, Any]:
        return {
            "tool_name": self.tool_name,
            "attempts": self.attempts,
            "distinct_consumers": self.distinct_consumers,
            "implemented": self.implemented,
            "stub": self.stub,
            "latent_demand": self.latent_demand,
            "last_ts": self.last_ts,
        }


def _connect_ro(db_path: Path) -> sqlite3.Connection:
    if not db_path.exists():
        raise FileNotFoundError(f"telemetry db not found: {db_path}")
    uri = f"file:{db_path.as_posix()}?mode=ro"
    return sqlite3.connect(uri, uri=True)


def _query_tools(conn: sqlite3.Connection) -> list[ToolRow]:
    rows = conn.execute(
        """
        SELECT
          tool_name,
          COUNT(*) AS attempts,
          COUNT(DISTINCT client_id_hash) AS distinct_consumers,
          SUM(CASE WHEN outcome = 'implemented' THEN 1 ELSE 0 END),
          SUM(CASE WHEN outcome = 'stub' THEN 1 ELSE 0 END),
          SUM(CASE WHEN outcome = 'latent_demand' THEN 1 ELSE 0 END),
          MAX(ts)
        FROM calls
        GROUP BY tool_name
        ORDER BY attempts DESC, tool_name ASC
        """
    ).fetchall()
    return [
        ToolRow(
            tool_name=r[0],
            attempts=r[1],
            distinct_consumers=r[2],
            implemented=r[3] or 0,
            stub=r[4] or 0,
            latent_demand=r[5] or 0,
            last_ts=r[6],
        )
        for r in rows
    ]


def _catalog_names() -> frozenset[str]:
    if not _CATALOG_AVAILABLE:
        return frozenset()
    try:
        return frozenset(e.name for e in load_catalog())
    except Exception:
        return frozenset()


def summarise(db_path: Path) -> dict[str, Any]:
    """Pure-function summary used by both the CLI and the tests."""
    conn = _connect_ro(db_path)
    try:
        tools = _query_tools(conn)
    finally:
        conn.close()
    catalog = _catalog_names()
    total_attempts = sum(t.attempts for t in tools)
    total_consumers = len({})  # placeholder; recomputed below
    consumer_set: set[str] = set()
    conn2 = _connect_ro(db_path)
    try:
        for (cid,) in conn2.execute(
            "SELECT DISTINCT client_id_hash FROM calls"
        ):
            consumer_set.add(cid)
    finally:
        conn2.close()
    total_consumers = len(consumer_set)
    latent_names = [
        t.tool_name for t in tools
        if t.latent_demand > 0
        and catalog
        and t.tool_name not in catalog
    ]
    return {
        "db_path": str(db_path),
        "total_attempts": total_attempts,
        "total_distinct_consumers": total_consumers,
        "tools": [t.as_dict() for t in tools],
        "latent_demand_names": sorted(set(latent_names)),
        "catalog_known": len(catalog) > 0,
    }


def _print_human(report: dict[str, Any]) -> None:
    print(
        f"📊  {report['total_attempts']} attempts across "
        f"{len(report['tools'])} tool(s) — "
        f"{report['total_distinct_consumers']} distinct consumer(s)"
    )
    print(f"   db: {report['db_path']}")
    print()
    if not report["tools"]:
        print("(no telemetry rows — run scripts/mcp_telemetry_store.py first)")
        return
    print(
        f"  {'tool':<28}  {'att':>5}  {'cons':>5}  "
        f"{'impl':>5}  {'stub':>5}  {'lat':>5}  last_ts"
    )
    for t in report["tools"]:
        print(
            f"  {t['tool_name']:<28}  {t['attempts']:>5}  "
            f"{t['distinct_consumers']:>5}  {t['implemented']:>5}  "
            f"{t['stub']:>5}  {t['latent_demand']:>5}  "
            f"{t['last_ts'] or '—'}"
        )
    if report["latent_demand_names"]:
        print()
        print("⚠️  latent-demand names not in catalog:")
        for n in report["latent_demand_names"]:
            print(f"   - {n}")


def _build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Query the MCP telemetry SQLite store (Phase 2 K2)."
    )
    parser.add_argument("--consumer-root", type=Path, default=None)
    parser.add_argument("--db", type=Path, default=None)
    parser.add_argument("--json", action="store_true")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = _build_parser().parse_args(argv)
    db_path = args.db or resolve_db(args.consumer_root)
    try:
        report = summarise(db_path)
    except FileNotFoundError as exc:
        msg = (
            f"❌  {exc}\n"
            "   run `python3 scripts/mcp_telemetry_store.py` first."
        )
        if args.json:
            print(json.dumps({"error": str(exc)}, separators=(",", ":")))
        else:
            print(msg, file=sys.stderr)
        return 1
    if args.json:
        print(json.dumps(report, separators=(",", ":")))
    else:
        _print_human(report)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/scripts/mcp_telemetry_query.ts
// MCP telemetry query CLI — Phase 2 K2.
//
// Reads the SQLite store written by `scripts/mcp_telemetry_store.ts`
// and surfaces:
//
// - Per-tool attempt counts.
// - Distinct-consumer counts (`client_id_hash`).
// - Outcome ratios (`implemented` / `stub` / `latent_demand`).
// - Latent-demand names — tool names not in the catalog.
//
// Refresh cadence: cheap enough to run on every `task mcp:report`
// invocation. stdlib-only; reads (never writes) the DB via Node's
// built-in `node:sqlite`, imported lazily.
//
// TS twin of mcp_telemetry_query.py (py2ts). Mirrors the full public
// surface: ToolRow, summarise, main.
import * as fs from 'node:fs';
import * as path from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import type { DatabaseSync } from 'node:sqlite';

import { resolve_db } from './mcp_telemetry_store.js';

export interface ToolRow {
    tool_name: string;
    attempts: number;
    distinct_consumers: number;
    implemented: number;
    stub: number;
    latent_demand: number;
    last_ts: string | null;
}

/** Field-ordered dict mirroring ToolRow.as_dict(). */
function toolRowAsDict(t: ToolRow): Record<string, unknown> {
    return {
        tool_name: t.tool_name,
        attempts: t.attempts,
        distinct_consumers: t.distinct_consumers,
        implemented: t.implemented,
        stub: t.stub,
        latent_demand: t.latent_demand,
        last_ts: t.last_ts,
    };
}

/** Python error subclass marker so main() can mirror the FileNotFoundError branch. */
class FileNotFoundLikeError extends Error {}

// Python's stdlib sqlite3 emits nothing on stderr; node:sqlite is flagged
// experimental on node 22 and prints an `ExperimentalWarning: SQLite …` line
// to stderr on first import (stable / silent on node >= 23). Drop only that
// specific warning so the twin's stderr stays byte-identical to the Python
// original across node versions. Installed once, narrowly matched.
let _sqliteWarningSilenced = false;
function _silenceSqliteExperimentalWarning(): void {
    if (_sqliteWarningSilenced) return;
    _sqliteWarningSilenced = true;
    const orig = process.emitWarning.bind(process);
    process.emitWarning = ((warning: string | Error, ...rest: unknown[]): void => {
        const text = typeof warning === 'string' ? warning : (warning?.message ?? '');
        if (/SQLite is an experimental/i.test(text)) return;
        (orig as (w: string | Error, ...a: unknown[]) => void)(warning, ...rest);
    }) as typeof process.emitWarning;
}

/** Lazy `node:sqlite` import — keeps the module loadable on a runtime without it. */
async function _loadSqlite(): Promise<typeof import('node:sqlite')> {
    _silenceSqliteExperimentalWarning();
    try {
        return await import('node:sqlite');
    } catch (exc) {
        const message = exc instanceof Error ? exc.message : String(exc);
        throw new Error(
            `node:sqlite is unavailable in this runtime (${message}). ` +
                'mcp_telemetry_query requires Node with the built-in SQLite module.',
        );
    }
}

/** Open read-only. Mirrors `_connect_ro`; raises FileNotFound-like on absence. */
async function _connect_ro(db_path: string): Promise<DatabaseSync> {
    if (!fs.existsSync(db_path)) {
        throw new FileNotFoundLikeError(`telemetry db not found: ${db_path}`);
    }
    const { DatabaseSync } = await _loadSqlite();
    // `file:<posix>?mode=ro` — Python opens via the same URI form.
    const uri = `file:${_asPosix(db_path)}?mode=ro`;
    return new DatabaseSync(uri, { readOnly: true });
}

/** Mirror Python `Path.as_posix()` — forward slashes on every platform. */
function _asPosix(p: string): string {
    return p.split(path.sep).join('/');
}

function _query_tools(conn: DatabaseSync): ToolRow[] {
    const rows = conn
        .prepare(
            `
        SELECT
          tool_name,
          COUNT(*) AS attempts,
          COUNT(DISTINCT client_id_hash) AS distinct_consumers,
          SUM(CASE WHEN outcome = 'implemented' THEN 1 ELSE 0 END),
          SUM(CASE WHEN outcome = 'stub' THEN 1 ELSE 0 END),
          SUM(CASE WHEN outcome = 'latent_demand' THEN 1 ELSE 0 END),
          MAX(ts)
        FROM calls
        GROUP BY tool_name
        ORDER BY attempts DESC, tool_name ASC
        `,
        )
        .all();
    return rows.map((r) => {
        // node:sqlite returns named columns; positional aggregates carry the
        // SQL expression as the key. Read columns in SELECT order to mirror
        // the Python positional tuple (r[0]..r[6]).
        const tuple = _rowValues(r);
        return {
            tool_name: tuple[0] as string,
            attempts: _asInt(tuple[1]),
            distinct_consumers: _asInt(tuple[2]),
            implemented: _asInt(tuple[3]) || 0,
            stub: _asInt(tuple[4]) || 0,
            latent_demand: _asInt(tuple[5]) || 0,
            last_ts: (tuple[6] ?? null) as string | null,
        };
    });
}

/** Values of a node:sqlite row object in column (insertion) order. */
function _rowValues(row: Record<string, unknown>): unknown[] {
    return Object.keys(row).map((k) => row[k]);
}

/** Coerce a SQLite numeric column to a JS number (bigint-safe). */
function _asInt(value: unknown): number {
    if (typeof value === 'bigint') {
        return Number(value);
    }
    if (typeof value === 'number') {
        return value;
    }
    if (value === null || value === undefined) {
        return 0;
    }
    return Number(value);
}

/** Catalog names — defensive: empty set if the catalog can't be loaded. */
async function _catalog_names(): Promise<Set<string>> {
    try {
        const mod = await import('./mcp_server/catalog.js');
        return new Set(mod.load_catalog().map((e) => e.name));
    } catch {
        return new Set();
    }
}

/** Pure-function summary used by both the CLI and the tests. */
export async function summarise(db_path: string): Promise<Record<string, unknown>> {
    let conn = await _connect_ro(db_path);
    let tools: ToolRow[];
    try {
        tools = _query_tools(conn);
    } finally {
        conn.close();
    }
    const catalog = await _catalog_names();
    const total_attempts = tools.reduce((acc, t) => acc + t.attempts, 0);
    const consumer_set = new Set<string>();
    conn = await _connect_ro(db_path);
    try {
        for (const row of conn.prepare('SELECT DISTINCT client_id_hash FROM calls').all()) {
            consumer_set.add(_rowValues(row)[0] as string);
        }
    } finally {
        conn.close();
    }
    const total_consumers = consumer_set.size;
    const latent_names = tools
        .filter((t) => t.latent_demand > 0 && catalog.size > 0 && !catalog.has(t.tool_name))
        .map((t) => t.tool_name);
    return {
        db_path,
        total_attempts,
        total_distinct_consumers: total_consumers,
        tools: tools.map((t) => toolRowAsDict(t)),
        latent_demand_names: _sortedUniqueByCodePoint(latent_names),
        catalog_known: catalog.size > 0,
    };
}

/** Python `sorted(set(names))` — dedupe then sort by Unicode code point. */
function _sortedUniqueByCodePoint(names: string[]): string[] {
    return [...new Set(names)].sort(_cmp_code_points);
}

function _cmp_code_points(a: string, b: string): number {
    const ai = [...a];
    const bi = [...b];
    const n = Math.min(ai.length, bi.length);
    for (let i = 0; i < n; i += 1) {
        const ca = (ai[i] as string).codePointAt(0) as number;
        const cb = (bi[i] as string).codePointAt(0) as number;
        if (ca !== cb) return ca - cb;
    }
    return ai.length - bi.length;
}

/** Unicode code-point length — Python `len(str)`, not UTF-16 `.length`. */
function _pyLen(s: string): number {
    return [...s].length;
}

/** Left-justify to `width` code points with spaces — Python `{:<width}`. */
function _ljust(s: string, width: number): string {
    const pad = width - _pyLen(s);
    return pad > 0 ? s + ' '.repeat(pad) : s;
}

/** Right-justify to `width` code points with spaces — Python `{:>width}`. */
function _rjust(value: number | string, width: number): string {
    const s = String(value);
    const pad = width - _pyLen(s);
    return pad > 0 ? ' '.repeat(pad) + s : s;
}

function _print_human(report: Record<string, unknown>): void {
    const tools = report['tools'] as Record<string, unknown>[];
    process.stdout.write(
        `📊  ${report['total_attempts']} attempts across ` +
            `${tools.length} tool(s) — ` +
            `${report['total_distinct_consumers']} distinct consumer(s)\n`,
    );
    process.stdout.write(`   db: ${report['db_path']}\n`);
    process.stdout.write('\n');
    if (tools.length === 0) {
        process.stdout.write('(no telemetry rows — run scripts/mcp_telemetry_store.py first)\n');
        return;
    }
    process.stdout.write(
        `  ${_ljust('tool', 28)}  ${_rjust('att', 5)}  ${_rjust('cons', 5)}  ` +
            `${_rjust('impl', 5)}  ${_rjust('stub', 5)}  ${_rjust('lat', 5)}  last_ts\n`,
    );
    for (const t of tools) {
        process.stdout.write(
            `  ${_ljust(t['tool_name'] as string, 28)}  ${_rjust(t['attempts'] as number, 5)}  ` +
                `${_rjust(t['distinct_consumers'] as number, 5)}  ${_rjust(t['implemented'] as number, 5)}  ` +
                `${_rjust(t['stub'] as number, 5)}  ${_rjust(t['latent_demand'] as number, 5)}  ` +
                `${(t['last_ts'] as string | null) ?? '—'}\n`,
        );
    }
    const latent = report['latent_demand_names'] as string[];
    if (latent.length > 0) {
        process.stdout.write('\n');
        process.stdout.write('⚠️  latent-demand names not in catalog:\n');
        for (const n of latent) {
            process.stdout.write(`   - ${n}\n`);
        }
    }
}

/**
 * Compact JSON, Python `json.dumps(obj, separators=(",", ":"))` byte-parity:
 * insertion order preserved, `ensure_ascii=True` (non-ASCII → `\uXXXX`).
 */
function _py_json_dumps(value: unknown): string {
    if (value === null) {
        return 'null';
    }
    switch (typeof value) {
        case 'boolean':
            return value ? 'true' : 'false';
        case 'number':
            return String(value);
        case 'string':
            return _py_json_string(value);
        case 'object':
            break;
        default:
            throw new TypeError(`Object of type ${typeof value} is not JSON serializable`);
    }
    if (Array.isArray(value)) {
        return `[${value.map((v) => _py_json_dumps(v)).join(',')}]`;
    }
    const obj = value as Record<string, unknown>;
    const parts = Object.keys(obj).map(
        (k) => `${_py_json_string(k)}:${_py_json_dumps(obj[k])}`,
    );
    return `{${parts.join(',')}}`;
}

/** Escape a string like Python `json.dumps(..., ensure_ascii=True)`. */
function _py_json_string(s: string): string {
    let out = '"';
    for (let i = 0; i < s.length; i += 1) {
        const code = s.charCodeAt(i);
        const ch = s[i] as string;
        if (ch === '"') out += '\\"';
        else if (ch === '\\') out += '\\\\';
        else if (ch === '\b') out += '\\b';
        else if (ch === '\f') out += '\\f';
        else if (ch === '\n') out += '\\n';
        else if (ch === '\r') out += '\\r';
        else if (ch === '\t') out += '\\t';
        else if (code >= 0x20 && code <= 0x7e) out += ch;
        else out += `\\u${code.toString(16).padStart(4, '0')}`;
    }
    return out + '"';
}

interface ParsedArgs {
    consumer_root: string | null;
    db: string | null;
    json: boolean;
}

/** Minimal argparse mirror for this CLI's flags. */
function _parse_args(argv: string[]): ParsedArgs {
    const args: ParsedArgs = { consumer_root: null, db: null, json: false };
    for (let i = 0; i < argv.length; i += 1) {
        const a = argv[i] as string;
        if (a === '--consumer-root') args.consumer_root = argv[++i] as string;
        else if (a.startsWith('--consumer-root=')) args.consumer_root = a.slice('--consumer-root='.length);
        else if (a === '--db') args.db = argv[++i] as string;
        else if (a.startsWith('--db=')) args.db = a.slice('--db='.length);
        else if (a === '--json') args.json = true;
    }
    return args;
}

export async function main(argv?: string[]): Promise<number> {
    const args = _parse_args(argv ?? process.argv.slice(2));
    const db_path = args.db ?? resolve_db(args.consumer_root);
    let report: Record<string, unknown>;
    try {
        report = await summarise(db_path);
    } catch (exc) {
        if (exc instanceof FileNotFoundLikeError) {
            const msg = `❌  ${exc.message}\n   run \`python3 scripts/mcp_telemetry_store.py\` first.`;
            if (args.json) {
                process.stdout.write(_py_json_dumps({ error: exc.message }) + '\n');
            } else {
                process.stderr.write(msg + '\n');
            }
            return 1;
        }
        throw exc;
    }
    if (args.json) {
        process.stdout.write(_py_json_dumps(report) + '\n');
    } else {
        _print_human(report);
    }
    return 0;
}

// Entry point — `process.exitCode`, never `process.exit`, so large stdout
// drains before the process ends.
const _HERE = fileURLToPath(import.meta.url);
const _isCliEntry =
    process.argv[1] !== undefined &&
    import.meta.url === pathToFileURL(path.resolve(process.argv[1])).href;
if (_isCliEntry || process.argv[1] === _HERE) {
    void main().then((rc) => {
        process.exitCode = rc;
    });
}
